doc/Class.3
'\"
'\" Generated from file 'Class\&.man' by tcllib/doctools with format 'nroff'
'\" Copyright (c) 2014-2019 Stefan Sobernig <stefan\&.sobernig@wu\&.ac\&.at>, Gustaf Neumann <gustaf\&.neumann@wu\&.ac\&.at>; available under the Creative Commons Attribution 3\&.0 Austria license (CC BY 3\&.0 AT)\&.
'\"
.TH "nx::Class" 3 2\&.5\&.0 Class "NX API"
.\" The -*- nroff -*- definitions below are for supplemental macros used
.\" in Tcl/Tk manual entries.
.\"
.\" .AP type name in/out ?indent?
.\"	Start paragraph describing an argument to a library procedure.
.\"	type is type of argument (int, etc.), in/out is either "in", "out",
.\"	or "in/out" to describe whether procedure reads or modifies arg,
.\"	and indent is equivalent to second arg of .IP (shouldn't ever be
.\"	needed;  use .AS below instead)
.\"
.\" .AS ?type? ?name?
.\"	Give maximum sizes of arguments for setting tab stops.  Type and
.\"	name are examples of largest possible arguments that will be passed
.\"	to .AP later.  If args are omitted, default tab stops are used.
.\"
.\" .BS
.\"	Start box enclosure.  From here until next .BE, everything will be
.\"	enclosed in one large box.
.\"
.\" .BE
.\"	End of box enclosure.
.\"
.\" .CS
.\"	Begin code excerpt.
.\"
.\" .CE
.\"	End code excerpt.
.\"
.\" .VS ?version? ?br?
.\"	Begin vertical sidebar, for use in marking newly-changed parts
.\"	of man pages.  The first argument is ignored and used for recording
.\"	the version when the .VS was added, so that the sidebars can be
.\"	found and removed when they reach a certain age.  If another argument
.\"	is present, then a line break is forced before starting the sidebar.
.\"
.\" .VE
.\"	End of vertical sidebar.
.\"
.\" .DS
.\"	Begin an indented unfilled display.
.\"
.\" .DE
.\"	End of indented unfilled display.
.\"
.\" .SO ?manpage?
.\"	Start of list of standard options for a Tk widget. The manpage
.\"	argument defines where to look up the standard options; if
.\"	omitted, defaults to "options". The options follow on successive
.\"	lines, in three columns separated by tabs.
.\"
.\" .SE
.\"	End of list of standard options for a Tk widget.
.\"
.\" .OP cmdName dbName dbClass
.\"	Start of description of a specific option.  cmdName gives the
.\"	option's name as specified in the class command, dbName gives
.\"	the option's name in the option database, and dbClass gives
.\"	the option's class in the option database.
.\"
.\" .UL arg1 arg2
.\"	Print arg1 underlined, then print arg2 normally.
.\"
.\" .QW arg1 ?arg2?
.\"	Print arg1 in quotes, then arg2 normally (for trailing punctuation).
.\"
.\" .PQ arg1 ?arg2?
.\"	Print an open parenthesis, arg1 in quotes, then arg2 normally
.\"	(for trailing punctuation) and then a closing parenthesis.
.\"
.\"	# Set up traps and other miscellaneous stuff for Tcl/Tk man pages.
.if t .wh -1.3i ^B
.nr ^l \n(.l
.ad b
.\"	# Start an argument description
.de AP
.ie !"\\$4"" .TP \\$4
.el \{\
.   ie !"\\$2"" .TP \\n()Cu
.   el          .TP 15
.\}
.ta \\n()Au \\n()Bu
.ie !"\\$3"" \{\
\&\\$1 \\fI\\$2\\fP (\\$3)
.\".b
.\}
.el \{\
.br
.ie !"\\$2"" \{\
\&\\$1	\\fI\\$2\\fP
.\}
.el \{\
\&\\fI\\$1\\fP
.\}
.\}
..
.\"	# define tabbing values for .AP
.de AS
.nr )A 10n
.if !"\\$1"" .nr )A \\w'\\$1'u+3n
.nr )B \\n()Au+15n
.\"
.if !"\\$2"" .nr )B \\w'\\$2'u+\\n()Au+3n
.nr )C \\n()Bu+\\w'(in/out)'u+2n
..
.AS Tcl_Interp Tcl_CreateInterp in/out
.\"	# BS - start boxed text
.\"	# ^y = starting y location
.\"	# ^b = 1
.de BS
.br
.mk ^y
.nr ^b 1u
.if n .nf
.if n .ti 0
.if n \l'\\n(.lu\(ul'
.if n .fi
..
.\"	# BE - end boxed text (draw box now)
.de BE
.nf
.ti 0
.mk ^t
.ie n \l'\\n(^lu\(ul'
.el \{\
.\"	Draw four-sided box normally, but don't draw top of
.\"	box if the box started on an earlier page.
.ie !\\n(^b-1 \{\
\h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.el \}\
\h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\l'|0u-1.5n\(ul'
.\}
.\}
.fi
.br
.nr ^b 0
..
.\"	# VS - start vertical sidebar
.\"	# ^Y = starting y location
.\"	# ^v = 1 (for troff;  for nroff this doesn't matter)
.de VS
.if !"\\$2"" .br
.mk ^Y
.ie n 'mc \s12\(br\s0
.el .nr ^v 1u
..
.\"	# VE - end of vertical sidebar
.de VE
.ie n 'mc
.el \{\
.ev 2
.nf
.ti 0
.mk ^t
\h'|\\n(^lu+3n'\L'|\\n(^Yu-1v\(bv'\v'\\n(^tu+1v-\\n(^Yu'\h'-|\\n(^lu+3n'
.sp -1
.fi
.ev
.\}
.nr ^v 0
..
.\"	# Special macro to handle page bottom:  finish off current
.\"	# box/sidebar if in box/sidebar mode, then invoked standard
.\"	# page bottom macro.
.de ^B
.ev 2
'ti 0
'nf
.mk ^t
.if \\n(^b \{\
.\"	Draw three-sided box if this is the box's first page,
.\"	draw two sides but no top otherwise.
.ie !\\n(^b-1 \h'-1.5n'\L'|\\n(^yu-1v'\l'\\n(^lu+3n\(ul'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.el \h'-1.5n'\L'|\\n(^yu-1v'\h'\\n(^lu+3n'\L'\\n(^tu+1v-\\n(^yu'\h'|0u'\c
.\}
.if \\n(^v \{\
.nr ^x \\n(^tu+1v-\\n(^Yu
\kx\h'-\\nxu'\h'|\\n(^lu+3n'\ky\L'-\\n(^xu'\v'\\n(^xu'\h'|0u'\c
.\}
.bp
'fi
.ev
.if \\n(^b \{\
.mk ^y
.nr ^b 2
.\}
.if \\n(^v \{\
.mk ^Y
.\}
..
.\"	# DS - begin display
.de DS
.RS
.nf
.sp
..
.\"	# DE - end display
.de DE
.fi
.RE
.sp
..
.\"	# SO - start of list of standard options
.de SO
'ie '\\$1'' .ds So \\fBoptions\\fR
'el .ds So \\fB\\$1\\fR
.SH "STANDARD OPTIONS"
.LP
.nf
.ta 5.5c 11c
.ft B
..
.\"	# SE - end of list of standard options
.de SE
.fi
.ft R
.LP
See the \\*(So manual entry for details on the standard options.
..
.\"	# OP - start of full description for a single option
.de OP
.LP
.nf
.ta 4c
Command-Line Name:	\\fB\\$1\\fR
Database Name:	\\fB\\$2\\fR
Database Class:	\\fB\\$3\\fR
.fi
.IP
..
.\"	# CS - begin code excerpt
.de CS
.RS
.nf
.ta .25i .5i .75i 1i
..
.\"	# CE - end code excerpt
.de CE
.fi
.RE
..
.\"	# UL - underline word
.de UL
\\$1\l'|0\(ul'\\$2
..
.\"	# QW - apply quotation marks to word
.de QW
.ie '\\*(lq'"' ``\\$1''\\$2
.\"" fix emacs highlighting
.el \\*(lq\\$1\\*(rq\\$2
..
.\"	# PQ - apply parens and quotation marks to word
.de PQ
.ie '\\*(lq'"' (``\\$1''\\$2)\\$3
.\"" fix emacs highlighting
.el (\\*(lq\\$1\\*(rq\\$2)\\$3
..
.\"	# QR - quoted range
.de QR
.ie '\\*(lq'"' ``\\$1''\\-``\\$2''\\$3
.\"" fix emacs highlighting
.el \\*(lq\\$1\\*(rq\\-\\*(lq\\$2\\*(rq\\$3
..
.\"	# MT - "empty" string
.de MT
.QW ""
..
.BS
.SH NAME
nx::Class \- API reference of the base-metaclass of the NX objectsystem
.SH SYNOPSIS
\fBnx::Class\fR \fBcreate\fR \fIcls\fR ?\fB-superclasses\fR \fIsuperClassNames\fR? ?\fB-mixins\fR \fImixinSpec\fR? ?\fB-filters\fR \fIfilterSpec\fR? ?\fIoption\fR \fIvalue\fR \&.\&.\&.? ?\fIinitBlock\fR?
.sp
\fBnx::Class\fR \fBnew\fR ?\fB-superclasses\fR \fIsuperClassNames\fR? ?\fB-mixins\fR \fImixinSpec\fR? ?\fB-filters\fR \fIfilterSpec\fR? ?\fIinitBlock\fR?
.sp
\fIcls\fR ?\fBpublic\fR | \fBprivate\fR | \fBprotected\fR? \fB alias\fR ?\fB-debug\fR? ?\fB-deprecated\fR? \fImethodName\fR ?\fB-returns\fR \fIvalueChecker\fR? ?\fB-frame\fR \fBobject\fR | \fBmethod\fR? \fIcmdName\fR
.sp
\fIcls\fR \fBcreate\fR \fIinstanceName\fR ?\fIoption\fR \fIvalue\fR \fIoption\fR \fIvalue\fR \&.\&.\&.?
.sp
\fIcls\fR \fBdelete\fR \fIfeature\fR \fIarg\fR
.sp
\fIcls\fR \fB\fR \fBfilters\fR \fIsubmethod\fR ?\fIarg\fR \&.\&.\&.?
.sp
\fIcls\fR ?\fBpublic\fR | \fBprotected\fR | \fBprivate\fR? \fB forward\fR ?\fB-debug\fR? ?\fB-deprecated\fR? \fImethodName\fR ?\fB-prefix\fR \fIprefixName\fR? ?\fB-frame\fR \fBobject\fR? ?\fB-returns\fR \fIvalueChecker\fR? ?\fB-verbose\fR? ?\fItarget\fR? ?\fIarg\fR \&.\&.\&.?
.sp
\fIcls\fR \fBinfo heritage\fR ?\fIpattern\fR?
.sp
\fIcls\fR \fBinfo instances\fR ?\fB-closure\fR? ?\fIpattern\fR?
.sp
\fIcls\fR \fBinfo mixinof\fR ?\fB-closure\fR? ?\fB-scope\fR \fIoption\fR? ?\fIpattern\fR?
.sp
\fIcls\fR \fBinfo subclasses\fR ?\fB-closure\fR? ?\fB-dependent\fR? ?\fIpattern\fR?
.sp
\fIcls\fR \fBinfo superclasses\fR ?\fB-closure\fR? ?\fIpattern\fR?
.sp
\fIcls\fR \fBinfo info\fR ?\fB-asList\fR?
.sp
\fIcls\fR \fBinfo  filters\fR ?\fB-guards\fR? ?\fIpattern\fR?
.sp
\fIcls\fR \fBinfo  method\fR \fIoption\fR \fImethodName\fR
.sp
\fIcls\fR \fBinfo  methods\fR ?\fB-callprotection\fR \fIlevel\fR? ?\fB-type\fR \fImethodType\fR? ?\fB-path\fR? ?\fInamePattern\fR?
.sp
\fIcls\fR \fBinfo  mixins\fR ?\fB-guards\fR? ?\fIpattern\fR?
.sp
\fIcls\fR \fBinfo  slots\fR ?\fB-type\fR \fIclassName\fR? ?\fIpattern\fR?
.sp
\fIcls\fR \fBinfo  variables\fR ?\fIpattern\fR?
.sp
\fIcls\fR ?\fBpublic\fR | \fBprotected\fR | \fBprivate\fR? \fB method\fR ?\fB-debug\fR? ?\fB-deprecated\fR? \fIname\fR \fIparameters\fR ?\fB-checkalways\fR? ?\fB-returns\fR \fIvalueChecker\fR? \fIbody\fR
.sp
\fIcls\fR \fB mixins\fR \fIsubmethod\fR ?\fIarg\fR \&.\&.\&.?
.sp
\fIcls\fR \fBnew\fR ?\fB-childof\fR \fIparentName\fR? ?\fIoption\fR \fIvalue\fR \fIoption\fR \fIvalue\fR \&.\&.\&.?
.sp
\fIcls\fR \fBproperty\fR ?\fB-accessor\fR \fBpublic\fR | \fBprotected\fR | \fBprivate\fR? ?\fB-class\fR \fIclassName\fR? ?\fB-configurable\fR \fItrueFalse\fR? ?\fB-incremental\fR? ?\fB-trace\fR \fBset\fR | \fBget\fR | \fBdefault\fR? \fIspec\fR ?\fIinitBlock\fR?
.sp
\fIcls\fR \fBrequire\fR ?\fBpublic\fR | \fBprotected\fR | \fBprivate\fR? \fB method\fR \fImethodName\fR
.sp
\fIcls\fR \fBvariable\fR ?\fB-accessor\fR \fBpublic\fR | \fBprotected\fR | \fBprivate\fR? ?\fB-incremental\fR? ?\fB-class\fR \fIclassName\fR? ?\fB-configurable\fR \fItrueFalse\fR? ?\fB-initblock\fR ?\fB-trace\fR \fBset\fR | \fBget\fR | \fBdefault\fR? \fIscript\fR? \fIspec\fR ?\fIdefaultValue\fR?
.sp
.BE
.SH DESCRIPTION
.PP
\fBnx::Class\fR is the base metaclass of the NX object
system\&. All classes (e\&.g\&. \fIcls\fR) are (direct or indirect)
instances of \fBnx::Class\fR\&. Therefore, the methods provided by \fBnx::Class\fR are
available to all classes\&. A class \fIcls\fR which does
not have \fBnx::Class\fR as its direct or indirect superclass is
referred to as an \fIapplication class\fR\&. By default, when
instantiating a new class from \fBnx::Class\fR, it becomes an
application class with \fBnx::Object\fR being set as its superclass\&. A
class \fIcls\fR which is explicitly declared as a (direct or
indirect) subclass of \fBnx::Class\fR is referred to as a \fImetaclass\fR, that
is, its instances will become classes as well\&. In other words, a
metaclass instantiates and subclasses \fBnx::Class\fR at the same
time\&.
.CS


+---------+
| ::nx::* |
+---------+--------------------------------------Y
|                                                |
|  instance of                                   |
|   \&.-------\&.                                    |
|  +--------'+     instance of     +----------+  |
|  |         |<\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.|          |  |
|  |  Class  |                     |  Object  |  |
|  |         |\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.>|          |  |
|  +---------+     subclass of     +-----+----+  |
|   ^   ^                                ^       |
\\\&.\&.\&.|\&.\&.\&.|\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.|\&.\&.\&.\&.\&.\&.\&./
    |   |                                |
    |   |subclass\&.\&.\&.\&.\&.(xor)\&.\&.\&.\&.\&.\&.subclass|
    |   |of       +-----------+        of|
    |   |\&.\&.\&.\&.\&.\&.\&.\&.\&.|           |\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.|
    | (metaclass) |   /cls/   | (application class)
    |\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.\&.|           |
    instance of   +-----------+

.CE
Classes can be created in the following ways:
.TP
\fBnx::Class\fR \fBcreate\fR \fIcls\fR ?\fB-superclasses\fR \fIsuperClassNames\fR? ?\fB-mixins\fR \fImixinSpec\fR? ?\fB-filters\fR \fIfilterSpec\fR? ?\fIoption\fR \fIvalue\fR \&.\&.\&.? ?\fIinitBlock\fR?
To create a class having the explicit name \fIcls\fR, use \fBcreate\fR\&.
.TP
\fBnx::Class\fR \fBnew\fR ?\fB-superclasses\fR \fIsuperClassNames\fR? ?\fB-mixins\fR \fImixinSpec\fR? ?\fB-filters\fR \fIfilterSpec\fR? ?\fIinitBlock\fR?
To create a class having an automatically assigned, implicit name, use \fBnew\fR\&.
.PP
The configuration options for direct and indirect instances of \fBnx::Class\fR,
which can be passed when calling \fBcreate\fR and \fBnew\fR, are
documented in the subsequent section\&.
.SH "CONFIGURATION OPTIONS FOR INSTANCES OF NX::CLASS"
.PP
Configuration options can be used for configuring objects during
their creation by passing the options as non-positional arguments into calls
of \fBnew\fR and \fBcreate\fR (see \fBnx::Class\fR)\&. An
existing object can be queried for its current configuration using
\fBcget\fR and it can be re-configured using \fBconfigure\fR\&.
.TP
\fB-superclasses\fR ?\fIsuperClassNames\fR?
If \fIsuperClassNames\fR is not specified, returns the superclasses of
the class\&. If provided, the class becomes the subclass of \fIsuperClassNames\fR\&.
.TP
\fB-filters\fR ?\fIfilterSpecs\fR?
Retrieves the list of filter methods currently active on instances of
the class, if \fIfilterSpecs\fR is not set\&. Otherwise, activates a
list of filter methods for the instances of the class\&. Filters are
returned or set in terms of a list of filter specifications\&.
.TP
\fB-mixins\fR ?\fImixinSpecs\fR?
Returns the list of mixin classes currently active on
instances of the class, if \fImixinSpecs\fR is not specified\&. Otherwise, the class
is extended by the list of mixin classes provided by \fImixinSpecs\fR\&.
mixin classes are returned or set in terms of a list of mixin specifications\&.
.PP
The configuration options provided by \fBnx::Object\fR are equally
available because an application class \fIcls\fR is an indirect
instance of \fBnx::Object\fR\&.
.SH "METHODS FOR INSTANCES OF NX::CLASS"
.TP
\fBalias\fR
.RS
.TP
\fIcls\fR ?\fBpublic\fR | \fBprivate\fR | \fBprotected\fR? \fB alias\fR ?\fB-debug\fR? ?\fB-deprecated\fR? \fImethodName\fR ?\fB-returns\fR \fIvalueChecker\fR? ?\fB-frame\fR \fBobject\fR | \fBmethod\fR? \fIcmdName\fR
Define an alias method for the given class\&. The
resulting method registers a pre-existing Tcl command \fIcmdName\fR
under the (alias) name \fImethodName\fR with the class\&. If \fIcmdName\fR refers
to another \fBmethod\fR, the corresponding argument
should be a valid method handle\&. If a Tcl command (e\&.g\&., a
\fBproc\fR), the argument should be a fully qualified Tcl command
name\&. If aliasing a subcommand (e\&.g\&., \fBarray exists\fR) of a Tcl namespace ensemble (e\&.g\&., \fBarray\fR), \fIcmdName\fR must hold the fully qualified subcommand name (and not the ensemble name of the subcommand)\&.
.sp
As for a regular \fBclass method\fR, \fB-returns\fR
allows for setting a value checker on the values returned by
the aliased command \fIcmdName\fR\&.
.sp
When creating an alias method for
a \fIC-implemented\fR Tcl command (i\&.e\&., command defined using the
Tcl/NX C-API), \fB-frame\fR sets the scope
for variable references used in the aliased command\&. If the provided
value is \fBobject\fR, then variable references will be resolved in the
context of the called object, i\&.e\&., the object upon which the alias method is invoked, as if they were object variables\&. There is no need for using
the colon-prefix notation for identifying object variables\&. If the
value is \fBmethod\fR, then the aliased command will be executed as a regular method call\&. The command is aware of its called-object context; i\&.e\&., it can resolve \fB::nx::self\fR\&. In addition, the alias method has access to the method-call context (e\&.g\&., \fBnx::next\fR)\&. If \fB-frame\fR is omitted, and by default, the variable references will resolve in the context of the caller of the alias method\&.
.sp
To express deprecation of the alias method \fImethodName\fR, set the \fB-deprecated\fR flag\&. Deprecated methods remain usable from client code, but their usage will be signaled to the developer and/or can be tracked using \fB::nsf::deprecated\fR\&. To register \fImethodName\fR with the debugger, set the \fB-debug\fR flag\&.  Entering and exiting a method, which was flagged for debugging, is recorded by calling the redefinable callback procs \fB::nsf::debug::call\fR and \fB::nsf::debug::exit\fR, respectively\&. By default, these callbacks forward to \fB::nsf::log\fR, which can also be customized at the script level\&.
.RE
.TP
\fB__class_configureparameter\fR
.RS
.TP
\fIcls\fR \fB__class_configureparameter\fR
Computes and returns the configuration options available for \fIcls\fR instances, to be consumed as method-parameter specification by \fBconfigure\fR\&.
.RE
.TP
\fBcreate\fR
.RS
.TP
\fIcls\fR \fBcreate\fR \fIinstanceName\fR ?\fIoption\fR \fIvalue\fR \fIoption\fR \fIvalue\fR \&.\&.\&.?
This factory method creates an instance \fIinstanceName\fR of \fIcls\fR
and returns \fIinstanceName\fR\&.
.CS


% nx::Class create AClass {
    :method init args {
      next
    }; # initialization method for instances of 'AClass'
  }; # defines a class 'AClass' being an instance of 'nx::Class'
::AClass
% ::AClass create anInstance; # defines an object 'anInstance' being an instance of 'AClass'
::anInstance
% ::anInstance info class
::AClass
% ::AClass info class
::nx::Class

.CE
.IP
\fBcreate\fR accepts the configuration options \fIoption\fR
available for this instance, such as those defined by properties of
\fIcls\fR (see \fBproperty\fR)\&.
.sp
Note that \fBcreate\fR is called internally when defining an
instance of \fIcls\fR using \fBnew\fR\&.
.sp
By calling \fBcreate\fR on \fBnx::Class\fR itself, the created
instance will become a new application class \fIinstanceName\fR on
which \fBcreate\fR can also be applied (i\&.e\&., it can be
instantiated)\&. If the so-created class has \fB::nx::Class\fR its
direct or indirect superclass, \fIinstanceName\fR is referred to as a
metaclass; that is, a class whose instances are again
classes\&.
.RE
.TP
\fBdelete\fR
.RS
.TP
\fIcls\fR \fBdelete\fR \fIfeature\fR \fIarg\fR
This method serves as the equivalent to Tcl's \fBrename\fR for
removing structural (properties, variables) and behavioral features
(methods) of the class:
.TP
\fIcls\fR \fBdelete  property\fR \fIpropertyName\fR
.TP
\fIcls\fR \fBdelete  variable\fR \fIvariableName\fR
.TP
\fIcls\fR \fBdelete  method\fR \fImethodName\fR
Removes a property \fIpropertyName\fR, variable \fIvariableName\fR,
and method \fImethodName\fR, respectively, previously defined for the
scope of the class\&.
.sp
\fBdelete  method\fR can be equally used for removing regular methods (see \fB method\fR), an alias method (see \fB alias\fR), and a forwarder method (see \fB forward\fR)\&.
.RE
.TP
\fBfilters\fR
.RS
.TP
\fIcls\fR \fB\fR \fBfilters\fR \fIsubmethod\fR ?\fIarg\fR \&.\&.\&.?
Accesses and modifies the list of methods which are registered as
filters with \fIcls\fR using a specific setter or getter
\fIsubmethod\fR:
.RS
.TP
\fIcls\fR \fB\fR \fBfilters add\fR \fIspec\fR ?\fIindex\fR?
Inserts a single filter into the current list of filters of \fIcls\fR\&. Using \fIindex\fR, a position in the existing list of filters for inserting the new filter can be set\&. If
omitted, \fIindex\fR defaults to the list head (0)\&.
.TP
\fIcls\fR \fB\fR \fBfilters clear\fR
Removes all filters from \fIcls\fR and returns the list of removed filters\&. Clearing
is equivalent to passing an empty list for \fIfilterSpecList\fR to
\fBclass\fR \fBfilter set\fR\&.
.TP

ChangeLog
2014-04-29 <Gustaf.Neumann@wu-wien.ac.at>
	* Release of XOTcl 1.6.8

2014-04-29 <Gustaf.Neumann@wu-wien.ac.at>
	* some more cleanup of autoconf (more quoting, remove obsolete file, use recent install-sh)
	* try to stick closer to current tcl conventions

2014-04-29 <stefan.sobernig@wu-wien.ac.at>
	* unix/tclAppInit.c: Fix one more compiler warning: undeclared Xotcl_Init()
	* Removing configure artifact, otherwise lintian complains
	* Makefile.in: LDFLAGS must be referenced differently under TEA
	* fixed "make install" problem, fix build problem for xowish

2014-04-28 <stefan.sobernig@wu-wien.ac.at>
	* library/xml/TclExpat-1.1/xmltok_impl.c: fix for CVE-2009-3720
	
2014-04-25 <Gustaf.Neumann@wu-wien.ac.at>
	* follow modern autoconf conventions
	* configure.a: use TEA 3.9
	* use new tcl.m4
	* Makefile.in: remove hard-coded "-rdynamic" from build of xotclsh
	and xowish
	* generic/xotclTrace.c: remove obsolete test

2014-04-24 <stefan.sobernig@wu-wien.ac.at>
	* Makefile.in: Make sure that xotclsh + xowish are explicitly linked
	against libtclstub*.so, rather than libtcl*.so. Otherwise, building
	with --with-xotclsh + --with-xowish fails starting with 8.6 (which
	effectively hides the various stub structures in libtcl*.so).
	* AppInit.c: Use the stubbed variant of Tcl_Init().
	* Tested under 8.6 (fossil trunk as of commit date) and 8.5 (fossil
	core-8-5-branch as of commit date).
	Thanks to Sergei Golovan (Debian Tcl/Tk Package Maintainers) for
	reporting the issue and an initial patch (see Debian bug #724816).

2013-07-16 <Gustaf.Neumann@wu-wien.ac.at>
	* library/actiweb/HttpPlace.xotcl: add URL query variables as
	arguments

2013-06-23 <Gustaf.Neumann@wu-wien.ac.at>
	* make version management simpler and freeze XOTcl 1.* versions In
	order to avoid bad interactions between XOTcl 1.0 and XOTcl 2.0
	the version dependency in 1.0 where changed to "package require
	-exact ... 1.0" where possible, and the provides where upgraded to
	1.0 in most cases
	* make sure that packages from XOTcl 1 require XOTcl 1
	* generic/xotcl.c: code cleanup

2013-03-26 <Gustaf.Neumann@wu-wien.ac.at>
	* doc/langRef.xotcl: update documentation

2012-10-13 <Gustaf.Neumann@wu-wien.ac.at>
	* Make sure to NS_EXPORT Ns_ModuleVersion for people using still
	the old-style aolserver module.

2012-01-16 <Gustaf.Neumann@wu-wien.ac.at>
	* xotcl.c Extend backport of handling of dashes in XOTcl's
	configure method to perform a more eager search for command
	begins. Extended regression test

2012-01-14 <Gustaf.Neumann@wu-wien.ac.at>
	* xotcl.c Don't overwrite error messages from __unknown handler in
	several situations (superclass, parameter class, mixin class)

2012-01-12 <Gustaf.Neumann@wu-wien.ac.at>
	* xotcl.c Backport from nsf: when processing arguments with
	leading dashes in "configure", accept only method names without
	spaces. This solves a problem with Tcl8.4 + ns_eval where the
	output of the serializer could not be processed by eval (a [list
	..] was lost).

2011-12-22 <Gustaf.Neumann@wu-wien.ac.at>
	* Httpd: force GMT dates as required by RFC

2011-11-12 <Gustaf.Neumann@wu-wien.ac.at>
	* Removed obsolete CVS-$Ids
	* Updated dates of Copyrights
	* Removed unneeded c++ hints for Emacs

2011-11-02 <Gustaf.Neumann@wu-wien.ac.at>
	* Release of XOTcl 1.6.7

2011-11-01 <Gustaf.Neumann@wu-wien.ac.at>
	* xotcl.c: some c-cleanup
	* library files: require exact version numbers to avoid conflicts
	  with XOTcl 2.0

2011-08-07 <Gustaf.Neumann@wu-wien.ac.at>
	* backport of autoname fix from nsf

2011-04-28 <Gustaf.Neumann@wu-wien.ac.at>
	* serializer: fix allChildren to handle names with semicolons
	* serializer: fix allInstances to handle names with semicolons

2011-02-09 <Gustaf.Neumann@wu-wien.ac.at>
	* Backport of nsf to handle partial implicit deletes

2011-01-20 <Gustaf.Neumann@wu-wien.ac.at>
	* Change backport of fix below to a check of numEntries

2011-01-20 <Gustaf.Neumann@wu-wien.ac.at>
	* Backport of fix for nsf : fixed a nasty bug within namespace
	deletion, when a deletion of one tcl cmd caused implicit deletions
	of other cmds in the same namespace. The classical idiom for
	looking over hash tables with Tcl_GetHashValue() and
	Tcl_NextHashEntry(hSrch) can lead to crashes (and has different
	behavior depending on the number of buckets). This fix handles
	complete implicit deletes.

2010-12-26 <Gustaf.Neumann@wu-wien.ac.at>
   *  Make cppcheck happy

2010-08-29 <Gustaf.Neumann@wu-wien.ac.at>
    * Fixed bug in filters reported by Kristoffer Lawson.  When a
      filter chain reaches its end without finding the invoked method,
      the unknown state was not resetted correctly and affected
      method invocations in the filter after "next".

2009-11-13 <Gustaf.Neumann@wu-wien.ac.at>
    * Release of XOTcl 1.6.6

2010-04-01 <Gustaf.Neumann@wu-wien.ac.at>
    * make sure, Tcl var traces are not called twice
      for volatile objects
    * fix for info: make sure, value is always initialized

2010-03-27 <Gustaf.Neumann@wu-wien.ac.at>
    * avoid cyclic dependencies with namespace imported command
      during final cleanup

2010-03-26 <Gustaf.Neumann@wu-wien.ac.at>
    * fixed 64bit problem in expat code
    * fixed memory leaks in info methods and forward (error case)

2010-03-22 <Gustaf.Neumann@wu-wien.ac.at>
    * clean up compilation when compile with assertions turned on
    * make sure to use xotcl 1.* when regression testing xotcl 1.*

2010-03-06 <Gustaf.Neumann@wu-wien.ac.at>
    * use installed expat when specified
      This option is needed fro Debian packaging rules.
      (Many thanks to Stefan Sobernig for this contribution)
    * cleanup of redundant section in configure.in
    * update to TEA 3.7

2010-02-09 <Gustaf.Neumann@wu-wien.ac.at>
    * provide compatibility with 8.6b1
    * bumped version number to 1.6.6

2010-12-29 <Gustaf.Neumann@wu-wien.ac.at>
    * fix example in tutorial

2009-11-13 <Gustaf.Neumann@wu-wien.ac.at>
    * Release of XOTcl 1.6.5

2009-11-06 <Gustaf.Neumann@wu-wien.ac.at>

library/nx/class.tcl
package provide nx::class 1.0

namespace eval ::nsf {
  array set ::nsf::methodDefiningMethod {
    class 1
  }
}

namespace eval ::nx {
  nx::Class eval {
    :public alias "class method" ::nx::Object::slot::__object::method

    :public alias "class alias" ::nx::Object::slot::__object::alias
    :public alias "class forward" ::nx::Object::slot::__object::forward
    #:public method "class forward" args {
    #  puts stderr "CLASS CMD: [self] [current method] [current args]"
    #  :public object forward {*}$args
    #}

    :public alias "class info" ::nx::Object::slot::__info

    :public method "class filter" args {
      set what filter
      switch [llength $args] {
	0 {return [::nsf::relation [::nsf::self] object-$what]}
	1 {return [::nsf::relation [::nsf::self] object-$what {*}$args]}
	default {return [::nx::Object::slot::$what [lindex $args 0] \
			     [::nsf::self] object-$what \
			     {*}[lrange $args 1 end]]
	}
      }
    }
    :public method "class mixin" args {
      set what mixin
      switch [llength $args] {
	0 {return [::nsf::relation [::nsf::self] object-$what]}
	1 {return [::nsf::relation [::nsf::self] object-$what {*}$args]}
	default {return [::nx::Object::slot::$what [lindex $args 0] \
			     [::nsf::self] object-$what \
			     {*}[lrange $args 1 end]]
	}
      }
    }
    :public alias "class filterguard" ::nsf::methods::object::filterguard
    :public alias "class mixinguard" ::nsf::methods::object::mixinguard
    
  }

  #
  # provide aliases for "class property" and "class variable"
  #
  ::nx::Class eval {
    :alias "class property" ::nsf::classes::nx::Object::property
    :alias "class variable" ::nsf::classes::nx::Object::variable
  }

  #
  # provide aliases for "class delete"
  #
  ::nx::Class eval {
    :alias "class delete property" ::nx::Object::slot::__delete::property
    :alias "class delete variable" ::nx::Object::slot::__delete::variable
    :alias "class delete method" ::nx::Object::slot::__delete::method
  }

  #
  # info redirector
  #
  ::nx::Class eval {
    :alias "class info" ::nx::Object::slot::__info
  }
  ######################################################################
  # Provide method "require"
  ######################################################################
  Object eval {
    #
    # method require, base cases
    #
    :method "require class method" {methodName} {
      ::nsf::method::require [::nsf::self] $methodName 1
      return [:info lookup method $methodName]
    }
    #
    # method require, public explicitly
    #
    :method "require public class method" {methodName} {
      set result [:require class method $methodName]
      ::nsf::method::property [self] $result call-protected false
      return $result
    }
    #
    # method require, protected explicitly
    #
    :method "require protected class method" {methodName} {
      set result [:require class method $methodName]
      ::nsf::method::property [self] $result call-protected true
      return $result
    }
    #
    # method require, private explicitly
    #
    :method "require private class method" {methodName} {
      set result [:require class method $methodName]
      ::nsf::method::property [self] $result call-private true
      return $result
    }
  }
}